# require-file-header

Require file header comments with configurable required tags.

## Rule Details

This rule enforces that every file begins with a header comment containing specified metadata tags. It supports JSDoc, block, and line comment styles.

## Options

This rule accepts an options object with the following properties:

| Option | Type | Default | Description |
|--------|------|---------|-------------|
| `style` | `"jsdoc"` \| `"block"` \| `"line"` | `"jsdoc"` | Comment style for the header |
| `requiredTags` | `string[]` | `["@file"]` | Tags required in the header |
| `template` | `string` | (style-based) | Custom template for autofix |
| `allowShebang` | `boolean` | `true` | Allow shebang (#!) before header |
| `maxLinesBeforeHeader` | `number` | `0` | Max empty lines before header |

## Examples

### Default (JSDoc style with @file required)

**Valid:**

```js
/**
 * @file User authentication module
 */
function authenticate() {}
```

**Invalid:**

```js
// Missing header
function authenticate() {}
```

```js
/**
 * Just a description without @file tag
 */
function authenticate() {}
```

### With Multiple Required Tags

```json
{
  "rules": {
    "lint-my-lines/require-file-header": ["error", {
      "requiredTags": ["@file", "@author", "@license"]
    }]
  }
}
```

**Valid:**

```js
/**
 * @file User module
 * @author Jane Doe
 * @license MIT
 */
```

**Invalid:**

```js
/**
 * @file User module
 */
// Missing @author and @license
```

### Block Comment Style

```json
{
  "rules": {
    "lint-my-lines/require-file-header": ["error", {
      "style": "block"
    }]
  }
}
```

**Valid:**

```js
/*
 * @file Configuration loader
 */
```

**Invalid (wrong style):**

```js
/**
 * @file Configuration loader
 */
// JSDoc style when block is expected
```

### Line Comment Style

```json
{
  "rules": {
    "lint-my-lines/require-file-header": ["error", {
      "style": "line",
      "requiredTags": ["@file", "@author"]
    }]
  }
}
```

**Valid:**

```js
// @file Utility functions
// @author John Smith
```

### With Shebang

Shebang lines (`#!/usr/bin/env node`) are allowed before the header by default:

**Valid:**

```js
#!/usr/bin/env node
/**
 * @file CLI entry point
 */
```

### Custom Template

```json
{
  "rules": {
    "lint-my-lines/require-file-header": ["error", {
      "template": "/**\n * @file {filename}\n * @author Your Name\n * @copyright {year} Company\n */"
    }]
  }
}
```

Template placeholders:
- `{filename}` - The base filename
- `{date}` - Current date (YYYY-MM-DD)
- `{year}` - Current year
- `{author}` - Placeholder `[Author]`

## When Not To Use It

- When files don't need header documentation
- When using a different documentation system
- When header comments are generated by build tools

## Related Rules

- [require-jsdoc](require-jsdoc.md) - Require JSDoc for functions
- [valid-jsdoc](valid-jsdoc.md) - Validate JSDoc content
